File: test/integration/fixtures/echo_server.py
#!/usr/bin/env python3
"""A minimal MCP echo server for testing mcp-hub."""
import json
import sys

TOOLS = [
    {
        "name": "echo",
        "description": "Echo back the input text",
        "inputSchema": {
            "type": "object",
            "properties": {
                "text": {"type": "string", "description": "Text to echo"},
            },
            "required": ["text"],
        },
    },
    {
        "name": "add",
        "description": "Add two numbers",
        "inputSchema": {
            "type": "object",
            "properties": {
                "a": {"type": "number", "description": "First number"},
                "b": {"type": "number", "description": "Second number"},
            },
            "required": ["a", "b"],
        },
    },
]


def handle_request(req):
    req_id = req.get("id")
    method = req.get("method")
    params = req.get("params", {})

    if method == "tools/list":
        return {"jsonrpc": "2.0", "id": req_id, "result": {"tools": TOOLS}}

    elif method == "tools/call":
        name = params.get("name", "")
        args = params.get("arguments", {})

        if name == "echo":
            text = args.get("text", "")
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "result": {
                    "content": [{"type": "text", "text": f"Echo: {text}"}],
                },
            }

        elif name == "add":
            a = args.get("a", 0)
            b = args.get("b", 0)
            result = a + b
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "result": {
                    "content": [{"type": "text", "text": str(result)}],
                },
            }

        else:
            return {
                "jsonrpc": "2.0",
                "id": req_id,
                "error": {"code": -32601, "message": f"Tool not found: {name}"},
            }

    else:
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "error": {"code": -32601, "message": f"Method not found: {method}"},
        }


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
            resp = handle_request(req)
            sys.stdout.write(json.dumps(resp) + "\n")
            sys.stdout.flush()
        except json.JSONDecodeError as e:
            sys.stderr.write(f"JSON parse error: {e}\n")
            sys.stderr.flush()


if __name__ == "__main__":
    main()

File: test/integration/fixtures/echo_server.js
#!/usr/bin/env node
const readline = require("node:readline");

const TOOLS = [
  {
    name: "echo",
    description: "Echo back the input text",
    inputSchema: {
      type: "object",
      properties: {
        text: { type: "string", description: "Text to echo" },
      },
      required: ["text"],
    },
  },
  {
    name: "add",
    description: "Add two numbers",
    inputSchema: {
      type: "object",
      properties: {
        a: { type: "number", description: "First number" },
        b: { type: "number", description: "Second number" },
      },
      required: ["a", "b"],
    },
  },
];

function handleRequest(req) {
  const reqId = req.id;
  const method = req.method;
  const params = req.params || {};

  if (method === "tools/list") {
    return { jsonrpc: "2.0", id: reqId, result: { tools: TOOLS } };
  }

  if (method === "tools/call") {
    const name = params.name || "";
    const args = params.arguments || {};

    if (name === "echo") {
      const text = args.text || "";
      return {
        jsonrpc: "2.0",
        id: reqId,
        result: { content: [{ type: "text", text: `Echo: ${text}` }] },
      };
    }

    if (name === "add") {
      const a = args.a || 0;
      const b = args.b || 0;
      return {
        jsonrpc: "2.0",
        id: reqId,
        result: { content: [{ type: "text", text: String(a + b) }] },
      };
    }

    return {
      jsonrpc: "2.0",
      id: reqId,
      error: { code: -32601, message: `Tool not found: ${name}` },
    };
  }

  return {
    jsonrpc: "2.0",
    id: reqId,
    error: { code: -32601, message: `Method not found: ${method}` },
  };
}

const rl = readline.createInterface({ input: process.stdin });
rl.on("line", (line) => {
  line = line.trim();
  if (!line) return;
  try {
    const req = JSON.parse(line);
    const resp = handleRequest(req);
    process.stdout.write(JSON.stringify(resp) + "\n");
  } catch (err) {
    process.stderr.write(`JSON parse error: ${err.message}\n`);
  }
});
